### PR TITLE
[BFT Testing] Wintermute attack integration test

### PR DESCRIPTION
**Pull request by yhassanzadeh13**
This PR implements wintermute attack integration testing. The test is passing every time the attack is conducted successfully. The test works as follows: 

**Setup:** we have a Flow network with 2 corrupted ENs, 3 corrupted VNs, one honest EN, and one honest VN, as well as the minimum required number of honest nodes for the remaining roles. The corrupted nodes are controlled by a `WintemuteAttackOrchestrator` (aka the orchestrator). Chunk alpha is set to 3, meaning each chunk is assigned to 3 VNs. Also, the minimum required approvals to seal a chunk is set to 2. Hence, each chunk is definitely assigned to at least two corrupted VNs, and their approval is enough to conduct the attack. 

**Scenario:** the orchestrator corruptes an execution result, dictates that to the corrupted execution nodes under its control. When the receipt of that corrupted result is incorporated in a block, the corrupted VNs assigned to the chunks of that result are requesting relevant chunk data packs. At this point, the orchestrator dictates those VNs to blindly approve the corrupted chunks. While the orchestrator mutes any chunk data pack response for corrupted chunks to the honest VN, as well as any result approval for the conflicting execution result with the corrupted result (that is generated by the sole honest execution node). 

The test is passing when enough approvals are witnessed for the corrupted chunks in the test network, as well as the block corresponding to the corrupted result is sealed.

### CODE REVIEW

File: integration/tests/lib/result_approval_state.go [hunk 1]
```diff
@@ -42,34 +42,60 @@ func (r *ResultApprovalState) Add(sender flow.Identifier, approval *flow.ResultA
...
+	approvals := r.WaitForTotalApprovalsFrom(t, flow.IdentifierList{verNodeID}, resultID, chunkIndex, 1)
+	if len(approvals) > 0 {
+		return approvals[0]
+	}
+	return nil
+}
+
+// WaitForTotalApprovalsFrom waits until "count" number of result approval for the given execution result id and
+// the chunk index is disseminated in the network from any subset of the given (verification) ids.
+// It returns the captured result approval.
```
Comment (jordanschalm) [inline integration/tests/lib/result_approval_state.go:54]:
```suggestion
// It returns the captured result approvals.
```

File: integration/tests/lib/result_approval_state.go [hunk 2]
```diff
@@ -42,34 +42,60 @@ func (r *ResultApprovalState) Add(sender flow.Identifier, approval *flow.ResultA
...
-			return false
-		}
-
-		for _, approval := range approvals {
-			if !bytes.Equal(approval.Body.ExecutionResultID[:], resultID[:]) {
-				continue // execution result IDs do not match
+		for _, verificationId := range verificationIds {
+			approvals, ok := r.resultApprovals[verificationId]
+			if !ok {
+				return false
```
Comment (jordanschalm) [inline integration/tests/lib/result_approval_state.go:72]:
```suggestion
				continue
```
We should still check for messages from other VNs in this case right?

File: integration/tests/bft/wintermute/wintermute_test.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+	for i := 0; i < len(corruptedResult.Chunks); i++ {
+		w.ApprovalState.WaitForTotalApprovalsFrom(w.T(),
+			w.corruptedVnIds,
+			corruptedResult.ID(),
+			uint64(i),
+			2)
+	}
+
+	// waits till victim block gets sealed.
+	w.BlockState.WaitForSealed(w.T(), victimBlock.Header.Height)
```
Comment (jordanschalm) [inline integration/tests/bft/wintermute/wintermute_test.go:53]: ⚠️  This is waiting until **a block** with the same height as the victim block is sealed. It doesn't actually check that the specific block we expect is sealed. 

In general, the `BlockState` tracking is making some assumptions which are no longer safe under Byzantine testing  conditions. [Example here](https://github.com/onflow/flow-go/blob/58f80f66b2119c330d9a0a7e8cf37757087581b1/integration/tests/lib/block_state.go#L66-L71).

For this specific case, we could use the Access API to verify `victimBlock` is sealed:
* assert latest sealed block height is > `victimBlock.Height`
* assert querying block by height returns `victimBlock`

File: insecure/wintermute/attackOrchestrator_test.go [hunk 4]
```diff
@@ -546,3 +546,197 @@ func TestWintermuteChunkResponseForCorruptedChunks(t *testing.T) {
...
+		Protocol:        insecure.Protocol_MULTICAST,
+		TargetNum:       3,
+		TargetIds:       unittest.IdentifierListFixture(10),
+		FlowProtocolEvent: unittest.ResultApprovalFixture(
+			unittest.WithExecutionResultID(originalResult.ID()),
+			unittest.WithChunk(0)),
+	}
+
+	approvalPassThrough := &sync.WaitGroup{}
+	approvalPassThrough.Add(1)
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply insecure/wintermute/attackOrchestrator_test.go]: https://github.com/onflow/flow-go/pull/2463/commits/0b47eb69f62a9d45e2779a58538ec1af01f4684b

File: integration/tests/bft/wintermute/wintermute_test.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+		w.ApprovalState.WaitForTotalApprovalsFrom(w.T(),
+			w.corruptedVnIds,
+			corruptedResult.ID(),
+			uint64(i),
+			2)
+	}
+
+	// waits till sealing height goes beyond the victim block
+	w.BlockState.WaitForSealed(w.T(), victimBlock.Header.Height)
+	// then checks querying victim block by height returns the victim block itself.
```
Comment (peterargue) [inline integration/tests/bft/wintermute/wintermute_test.go:54]:
```suggestion
	// then checks querying victim block by height returns the original victim block.
```

**Review by peterargue — APPROVED** [review]
looks good. added a few small comments

File: integration/tests/bft/wintermute/wintermute_test.go [hunk 6]
```diff
@@ -49,8 +49,12 @@ func (w *WintermuteTestSuite) TestWintermuteAttack() {
 			2)
 	}
 
-	// waits till victim block gets sealed.
+	// waits till sealing height goes beyond the victim block
```
Comment (jordanschalm) [inline integration/tests/bft/wintermute/wintermute_test.go:52]:
```suggestion
	// waits until we seal a height equal to the victim block height
```

File: integration/tests/common/sealing_and_verification.go [hunk 7]
```diff
@@ -65,7 +66,13 @@ func SealingAndVerificationHappyPathTest(
 	}
 
 	// waits until blockB is sealed by consensus nodes after result approvals for all of its chunks emitted.
-	blockState.WaitForSealed(t, blockB.Header.Height)
+	// blockState.WaitForSealed(t, blockB.Header.Height)
+	// waits till sealing height goes beyond the victim block
+	blockState.WaitForSealed(t, blockB.Header.Height+1)
```
Comment (jordanschalm) [inline integration/tests/common/sealing_and_verification.go:71]:
```suggestion
	// waits until we seal a height equal to the victim block height
	blockState.WaitForSealed(t, blockB.Header.Height)
```
consistency with other test case

**Review by jordanschalm — APPROVED** [review]
🎸

File: insecure/wintermute/attackOrchestrator_test.go [hunk 8]
```diff
@@ -546,3 +546,194 @@ func TestWintermuteChunkResponseForCorruptedChunks(t *testing.T) {
...
+
+	// waits till miscellaneous event is sent to orchestrator.
+	unittest.RequireReturnsBefore(t,
+		sendMiscellaneousEventWG.Wait,
+		1*time.Second,
+		"could not send miscellaneous event on time to orchestrator")
+
+	// waits till miscellaneous event is passed through by the orchestrator.
+	unittest.RequireReturnsBefore(t,
+		sendMiscellaneousEventWG.Wait,
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply insecure/wintermute/attackOrchestrator_test.go]: https://github.com/onflow/flow-go/pull/2463/commits/d69fcb1d09263013973595ecf82108a49bf21e76

**Pull request merged**